pkg/cmd/roachtest/tests/versionupgrade.go

```diff
@@ -287,3 +290,69 @@ for i in 1 2 3 4; do
 done
 `)
 }
+
+func registerHTTPRestart(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:             "http-restart/mixed-version",
+		Owner:            registry.OwnerTestEng,
+		Cluster:          r.MakeClusterSpec(4),
+		CompatibleClouds: registry.AllClouds,
+		Suites:           registry.Suites(registry.MixedVersion, registry.Nightly),
+		Randomized:       true,
+		Run:              runHTTPRestart,
+		Timeout:          1 * time.Hour,
+	})
+}
+
+func runHTTPRestart(ctx context.Context, t test.Test, c cluster.Cluster) {
+	mvt := mixedversion.NewTest(ctx, t, t.L(), c,
+		c.CRDBNodes(),
+		mixedversion.MinimumSupportedVersion("v24.2.0"),
+		mixedversion.MinUpgrades(2),
+		mixedversion.EnabledDeploymentModes(mixedversion.SeparateProcessDeployment),
+	)
+
+	httpCall := func(ctx context.Context, node int, l *logger.Logger) {
+		logEvery := roachtestutil.Every(1 * time.Second)
+		adminURLs, err := c.ExternalAdminUIAddr(ctx, l, c.Node(node))
+		if err != nil {
+			t.Fatal(err)
+		}
+		url := "https://" + adminURLs[0] + "/ts/query"
+		l.Printf("Sending requests to %s", url)
+		client := roachtestutil.DefaultHTTPClient(c, l)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// Copied from elsewhere, actual HTTP req shouldn't matter
+			var response tspb.TimeSeriesQueryResponse
+			request := tspb.TimeSeriesQueryRequest{
+				StartNanos: timeutil.Now().UnixNano() - 10*time.Second.Nanoseconds(),
+				EndNanos:   timeutil.Now().UnixNano(),
+				// Ask for 10s intervals.
+				SampleNanos: (10 * time.Second).Nanoseconds(),
+				Queries: []tspb.Query{{
+					Name:             "cr.node.sql.service.latency-p90",
+					SourceAggregator: tspb.TimeSeriesQueryAggregator_MAX.Enum(),
+				}},
+			}
+			if err := client.PostProtobuf(ctx, url, &request, &response); err != nil {
+				if logEvery.ShouldLog() {
+					l.Printf("Error posting protobuf: %s", err)
+				}
+			}
+		}
+	}
+
+	for i := 1; i <= 4; i++ {
+		mvt.BackgroundFunc("HTTP requests", func(ctx context.Context, l *logger.Logger, rng *rand.Rand, h *mixedversion.Helper) error {
```

Comment:
Why not `OnStartup`? Its execution always precedes any background func. (See `testPlanner.testStartSteps`.)

Reply:
`BackgroundFunc` runs in a goroutine managed by framework. The framework knows to not block on these steps and expects them to run for the duration of the test. `OnStartup` is a normal step so the test will block until the step is finished. If you wanted to achieve something similar in `OnStartup` you'd have to add your own concurrency control with `h.GoWithCancel()`.

To answer Stan's question, there are no startup steps here so the ordering of the hooks is the exact same. So I just went with `BackgroundFunc` since the ergonomics of it are slightly cleaner imo, no strong preference though. 

Ideally it should run before any nodes are started, that way we don't miss out on testing the first time it starts up. There's currently no way to do that in framework though 😞

Reply:
> Ideally it should run before any nodes are started, that way we don't miss out on testing the first time it starts up. There's currently no way to do that in framework though.

Right, that was my thinking too...`BackgroundFunc` may not be able to catch a regression on _every_ run. I suppose that's a reasonable compromise for now. I wonder if we should create a follow-up issue to add the missing feature? To my knowledge, this is the first such use-case where we do want to execute a user-defined func. immediately at startup.

Reply:
I think the timing to pull that off makes it complicated. `c.Start` handles both the selecting the `adminUIPorts` and starting CRDB, but we'd need to interleave this hook between the two steps. A workaround would be to hardcode the ports used, but that also isn't supported out of the box since mixed version framework controls the creation of clusters/tenants.

Reply:
Yep. We'd need to refactor a bit. E.g., `c.Start` will take an optional callback func; it's called immediately after we know the port(s), i.e., after `generateStartArgs`. Then, we expose a new user hook inside the mixedversion framework; it gets invoked before the crdb process is started. It seems rather contrived to satisfy this particular use-case, but some other use-case will probably come up in the future. Let's file it away in a new issue so that we don't have to fish for it; worst-case it will linger like many others from ages ago :)

Reply:
Will do 👍